platform/src/ActivityValidator.js

```diff
@@ -0,0 +1,208 @@
+import { ConfigValidationError } from './ConfigValidationError';
+
+const ERROR_CATEGORY = "ERROR_CONFIG_DEF"
+
+const errorFileType =  {
+    ACTIVITY: "ActivityConfig",
+    TOOL: "ToolConfig"
+}
+
+class ActivityValidator {
+
+    // Activity checks
+    
+    /**
+     * Check layout panel ids exist
+     * @param {Object} activity - An activity with all references resolved.
+     * @returns {ConfigValidationError[]} Array of configuration errors, empty if there are none.
+     */
+    static checkLayoutPanelIdsExist(activity){
+        let errors = [];
+
+        let combinedRows = [];
+        activity.layout.area.forEach( (row) => combinedRows= combinedRows.concat(row) );
+
+        const ids = new Set(combinedRows);
+        ids.delete(""); // Remove empty references used to indicate extend panel
+
+        ids.forEach( (id) => {
+            if (!this.idExists(activity.panels, id)){
+                    errors.push( 
+                        new ConfigValidationError(ERROR_CATEGORY, "A panel does not exist for the given id.",  
+                        `${activity.id} -> layout -> ${id}`, errorFileType.ACTIVITY)
+                );
+            }
+        });
+
+        return errors;
+    }
+
+    /**
+     * Check the actions reference ids
+     * @param {*} activity - An activity with all references resolved.
+     * @returns {ConfigValidationError[]} Array of configuration errors, empty if there are none.
+     */
+    static checkActions(activity){
+        let errors = [];
+
+        activity.actions.forEach( (act) => {
+            errors = errors.concat( this.checkActionIdsExist(act, activity.panels) );
+        });
+
+        return errors;
+    }
+
+    /**
+     * Check an action panel ids exist
+     * @param {*} action - The action to check
+     * @param {*} panels - The panels 
+     * @returns {ConfigValidationError[]} Array of configuration errors, empty if there are none.
+     */
+    static checkActionIdsExist(action, panels){
+        let errors = [];
+
+        if (!this.idExists( panels, action.source.id)){ // action ids are resolved so get id from object
+            errors.push( 
+                new ConfigValidationError(ERROR_CATEGORY, "A panel does not exist for the given id.",  
+                `actions -> source: ${action.source}`, errorFileType.ACTIVITY)
+            );
+        }
+
+        if (!this.idExists( panels, action.output.id)){ // action ids are resolved so get id from object
+            errors.push( 
+                new ConfigValidationError(ERROR_CATEGORY, "A panel does not exist for the given id.",  
+                `actions -> output: ${action.output}`, errorFileType.ACTIVITY)
```

Comment:
Not `${action.output.id}`?

Reply:
As per https://github.com/mdenet/educationplatform/pull/118/files/dcc62b0aecc8468d790dbaba55466bc7792aab05#r1406005085

Reply:
Fixed in 40616a5525ed9b6fb0d7b2dc4de5fc691599691c